zipline/pipeline/loaders/earnings.py

```diff
@@ -0,0 +1,139 @@
+            EarningsCalendar.next_announcement,
+            next_earnings_date_frame(
+                self._all_dates,
+                self._announcement_dates,
+            ),
+            adjustments=None,
+        )
+
+    @lazyval
+    def previous_announcement_loader(self):
+        return DataFrameLoader(
+            EarningsCalendar.previous_announcement,
+            previous_earnings_date_frame(
+                self._all_dates,
+                self._announcement_dates,
+            ),
+            adjustments=None,
+        )
+
+    def load_adjusted_array(self, columns, dates, assets, mask):
+        return {
+            column: self.get_loader(column).load_adjusted_array(
+                [column], dates, assets, mask
+            )
+            for column in columns
+        }
+
+
+def next_earnings_date_frame(dates, announcement_dates):
+    """
+    Make a DataFrame representing simulated next earnings dates.
+
+    Parameters
+    ----------
+    dates : DatetimeIndex.
+        The index of the returned DataFrame.
+    announcement_dates : dict[int -> DatetimeIndex]
+        Dict mapping sids to an index of dates on which earnings were announced
+        for that sid.
+
+    Returns
+    -------
+    next_earnings: pd.DataFrame
+        A DataFrame representing, for each (label, date) pair, the first entry
+        in `earnings_calendars[label]` on or after `date`.  Entries falling
+        after the last date in a calendar will have `NaT` as the result in the
+        output.
+
+    See Also
+    --------
+    next_earnings_date_frame
+    """
+    cols = {equity: full_like(dates, "NaT") for equity in announcement_dates}
+    for equity, earnings_dates in iteritems(announcement_dates):
+        next_dt_indices = earnings_dates.searchsorted(dates)
+        mask = next_dt_indices < len(earnings_dates)
+        cols[equity][mask] = earnings_dates[next_dt_indices[mask]]
+
+    return pd.DataFrame(index=dates, data=cols)
+
+
+def previous_earnings_date_frame(dates, announcement_dates):
+    """
+    Make a DataFrame representing simulated next earnings dates.
+
+    Parameters
+    ----------
+    dates : DatetimeIndex.
+        The index of the returned DataFrame.
+    announcement_dates : dict[int -> DatetimeIndex]
+        Dict mapping sids to an index of dates on which earnings were announced
+        for that sid.
+
+    Returns
+    -------
+    prev_earnings: pd.DataFrame
+        A DataFrame representing, for (label, date) pair, the first entry in
+        `announcement_dates[label]` strictly before `date`.  Entries falling
+        before the first date in a calendar will have `NaT` as the result in
+        the output.
+
+    See Also
+    --------
+    next_earnings_date_frame
+    """
+    cols = {equity: full_like(dates, "NaT") for equity in announcement_dates}
+    for equity, earnings_dates in iteritems(announcement_dates):
+        # Subtract one to roll back to the index of the previous date.
```

Comment:
This is probably incorrect.  We shouldn't be subtracting on days where the earnings announcement date falls on the same day as the calendar date.